### ISSUE TITLE
"[.data.table" for an S4 class inheriting from "data.table"

### ISSUE THREAD

**Issue by otoomet** [Rdatatable/data.table#1881]
I am perhaps repeating #1504 here but just to start from a clean sheet.

`[.data.table` cannot handle S4 objects inheriting from `data.table`:

```
setClass("Foo", contains="data.table")
a <- new("Foo", data.table(a=1:2))
a[1,]
```

> Error in if (n > 0) c(NA_integer_, -n) else integer() : 
  argument is of length zero

`[.data.table` seems to be dispatched:

```
traceback()
# 4: .set_row_names(nrow(ans))
# 3: setattr(ans, "row.names", .set_row_names(nrow(ans)))
# 2: `[.data.table`(a, 1, )
# 1: a[1, ]
```

Note that it works for data.frames:

```
setClass("Bar", contains="data.frame")
a <- new("Bar", data.frame(a=1:2))
a[1,]
# [1] 1
```

Is there any hope this issue will be fixed in the near future?  It would make it easier to use data.table as a container for S4 objects.

Edit: this happens both with 1.9.6 and the current github version

**Comment by otoomet** [comment]
The issue seems to be related to the line
```setattr(ans, "class", class(x))```
(around line 1403 in R/data.table.R, in function "[.data.table".)  One cannot create S4 objects like this and hence ans will not be a class with valid rows, and hence the error when setting row.names below.  A quick hack that works for me right now is to replace the line with:
```
            if(isS4(x)) {
               ans <- new(class(x), ans)
            }
            else {
               setattr(ans, "class", class(x)) # fix for #5296
            }
```

There are two issues with this approach though:
* I suspect "new()" will copy the object, and hence contradict much of the data.table philosophy (have to check it though)
* it is not obvious that a subset of x should still be of the same class.  (I think it is more legitimate for [.data.table() always to create plain "data.table" objects.)

**Comment by otoomet** [comment]
I think '[.data.table' should return an object of class 'data.table' (or vector/list), not the of the original subclass.  Hence the code above should be
```
            if(isS4(x)) {
               ans <- as.data.table(ans)
            }
            else {
               setattr(ans, "class", class(x)) # fix for #5296
            }
```
This is because superclass methods cannot know if the result is still a valid subclass object.  Maybe S3 classes should be handled differently.

There seems to be places in the code that do not play well with S4 objects, for instance I was not able to transform the example above by ```as.data.table(a)```, leading again to problems related to classes.

**Comment by MichaelChirico** [comment]
Not an error on current version; closing

```
setClass("Foo", contains="data.table")
a <- new("Foo", data.table(a=1:2))
a[1,]
```
